File: engine/trades.py
from __future__ import annotations

import math

from engine.live import (
    get_live_display_return_point_at_or_before,
    get_live_return_point_at_or_before,
)
from engine.math_utils import corrcoef, nan_max, nan_mean, nan_median, nan_min, nan_percentile, nan_std
from engine.models import AssetMeta, EventReturns, SharedLiveSnapshot
from engine.returns import poi_ret, unit_label


def stars(iqr: float, median: float) -> str:
    if abs(median) < 0.01:
        return "....."
    ratio = iqr / (abs(median) + 1e-9)
    if ratio < 0.4:
        return "*****"
    if ratio < 0.7:
        return "****."
    if ratio < 1.0:
        return "***.."
    if ratio < 1.5:
        return "**..."
    return "*...."


def status_from_pctile(pctile: float | None) -> str:
    if pctile is None or math.isnan(pctile):
        return "--"
    if pctile < 25:
        return "Still open"
    if pctile < 50:
        return "On track"
    if pctile < 75:
        return "Chasing"
    return "Extended"


def compute_trade_rows(
    labels: list[str],
    event_returns: EventReturns,
    asset_meta: dict[str, AssetMeta],
    selected_events: list[str],
    day_n: int,
    fwd_days: int,
    live: SharedLiveSnapshot | None,
) -> list[dict]:
    forward_offset = day_n + fwd_days
    if forward_offset <= day_n:
        return []
    rows: list[dict] = []
    for label in labels:
        fwd_vals: list[float] = []
        for event_name in selected_events:
            start = poi_ret(event_returns, label, event_name, day_n)
            finish = poi_ret(event_returns, label, event_name, forward_offset)
            if not math.isnan(start) and not math.isnan(finish):
                fwd_vals.append(finish - start)
        if len(fwd_vals) < 2:
            continue
        median = nan_median(fwd_vals)
        mean = nan_mean(fwd_vals)
        std = nan_std(fwd_vals)
        iqr = nan_percentile(fwd_vals, 75) - nan_percentile(fwd_vals, 25)
        is_rates = asset_meta.get(label).is_rates_bp if label in asset_meta else False
        direction = 1 if median >= 0 else -1
        hit_rate = len([value for value in fwd_vals if value * direction > 0]) / len(fwd_vals)
        adjusted = [value * direction for value in fwd_vals]
        mean_adjusted = nan_mean(adjusted)
        std_adjusted = nan_std(adjusted)
        sharpe = mean_adjusted / (std_adjusted + 1e-9)
        downside = [value for value in adjusted if value < 0]
        downside_std = nan_std(downside) if len(downside) > 1 else std_adjusted + 1e-9
        sortino = mean_adjusted / (downside_std + 1e-9)
        worst = nan_min(fwd_vals) if direction > 0 else nan_max(fwd_vals)
        best = nan_max(fwd_vals) if direction > 0 else nan_min(fwd_vals)
        skew = 0.0
        if len(fwd_vals) >= 3 and std > 0:
            skew = sum(((value - mean) / std) ** 3 for value in fwd_vals) / len(fwd_vals)
        live_gap = math.nan
        live_pctile = math.nan
        live_point = get_live_return_point_at_or_before(live, label, day_n)
        if live_point:
            historical_at_live = [
                poi_ret(event_returns, label, event_name, live_point[0])
                for event_name in selected_events
            ]
            historical_at_live = [value for value in historical_at_live if not math.isnan(value)]
            if len(historical_at_live) >= 2:
                live_gap = direction * (live_point[1] - nan_median(historical_at_live))
                live_pctile = len([value for value in historical_at_live if live_point[1] > value]) / len(historical_at_live) * 100
        meta = asset_meta[label]
        rows.append(
            {
                "lbl": label,
                "cls": meta.class_name,
                "ticker": meta.ticker,
                "dir": "LONG" if median >= 0 else "SHORT",
                "med": median,
                "mean": mean,
                "std": std,
                "iqr": iqr,
                "stars": stars(iqr, median),
                "n": len(fwd_vals),
                "n_total": len(selected_events),
                "unit": unit_label(meta),
                "is_rates": is_rates,
                "hit_rate": hit_rate,
                "sharpe": sharpe,
                "sortino": sortino,
                "skew": skew,
                "worst": worst,
                "best": best,
                "live_gap": live_gap,
                "live_pctile": live_pctile,
                "status": status_from_pctile(None if math.isnan(live_pctile) else live_pctile),
                "fwd_vals": fwd_vals,
            }
        )
    return sorted(rows, key=lambda row: row["sharpe"], reverse=True)


def compute_per_horizon_stats(
    label: str,
    selected_events: list[str],
    day_n: int,
    event_returns: EventReturns,
    horizons: list[dict],
) -> list[dict]:
    rows: list[dict] = []
    for horizon in horizons:
        offset = horizon["offset"]
        if offset <= day_n:
            continue
        vals: list[float] = []
        for event_name in selected_events:
            start = poi_ret(event_returns, label, event_name, day_n)
            finish = poi_ret(event_returns, label, event_name, offset)
            if not math.isnan(start) and not math.isnan(finish):
                vals.append(finish - start)
        if len(vals) < 2:
            continue
        median = nan_median(vals)
        mean = nan_mean(vals)
        std = nan_std(vals)
        downside = [value for value in vals if value < 0]
        downside_std = nan_std(downside) if len(downside) > 1 else std + 1e-9
        direction = 1 if median >= 0 else -1
        adjusted = [value * direction for value in vals]
        rows.append(
            {
                "horizon_label": horizon["label"],
                "horizon_offset": offset,
                "q1": nan_percentile(vals, 25),
                "med": median,
                "q3": nan_percentile(vals, 75),
                "hit": len([value for value in vals if value * direction > 0]) / len(vals),
                "sharpe": (mean * direction) / (std + 1e-9),
                "sortino": (mean * direction) / (downside_std + 1e-9),
                "skew": 0 if len(vals) < 3 else sum(((value - mean) / (std + 1e-9)) ** 3 for value in vals) / len(vals),
                "worst": nan_min(adjusted),
                "n": len(vals),
            }
        )
    return rows


def build_dot_plot(label: str, selected_events: list[str], start_offset: int, end_offset: int, event_returns: EventReturns) -> list[dict]:
    points: list[dict] = []
    if end_offset <= start_offset:
        return points
    for event_name in selected_events:
        start = poi_ret(event_returns, label, event_name, start_offset)
        finish = poi_ret(event_returns, label, event_name, end_offset)
        if not math.isnan(start) and not math.isnan(finish):
            points.append({"event": event_name, "value": finish - start})
    return points


def build_live_deviation_series(label: str, selected_events: list[str], day_n: int, event_returns: EventReturns, live: SharedLiveSnapshot | None) -> list[dict]:
    series: list[dict] = []
    for offset in range(day_n + 1):
        values = [poi_ret(event_returns, label, event_name, offset) for event_name in selected_events]
        values = [value for value in values if not math.isnan(value)]
        live_point = get_live_display_return_point_at_or_before(live, label, offset)
        series.append(
            {
                "offset": offset,
                "p25": nan_percentile(values, 25) if values else None,
                "p75": nan_percentile(values, 75) if values else None,
                "median": nan_median(values) if values else None,
                "live": live_point[1] if live_point else None,
            }
        )
    return series


def build_idea_correlation_matrix(rows: list[dict], selected_events: list[str], day_n: int, fwd_days: int, event_returns: EventReturns) -> dict | None:
    focus_rows = rows[: min(12, len(rows))]
    if len(focus_rows) < 2:
        return None
    end_offset = day_n + fwd_days
    per_asset_values = []
    for row in focus_rows:
        values = []
        for event_name in selected_events:
            start = poi_ret(event_returns, row["lbl"], event_name, day_n)
            finish = poi_ret(event_returns, row["lbl"], event_name, end_offset)
            values.append(finish - start if not math.isnan(start) and not math.isnan(finish) else math.nan)
        per_asset_values.append(values)
    overlap_counts = []
    matrix = []
    for left_values in per_asset_values:
        overlap_row = []
        corr_row = []
        for right_values in per_asset_values:
            filtered_left = []
            filtered_right = []
            for left, right in zip(left_values, right_values):
                if not math.isnan(left) and not math.isnan(right):
                    filtered_left.append(left)
                    filtered_right.append(right)
            overlap_row.append(len(filtered_left))
            corr_row.append(corrcoef(filtered_left, filtered_right) if len(filtered_left) >= 3 else math.nan)
        overlap_counts.append(overlap_row)
        matrix.append(corr_row)
    return {"labels": [row["lbl"] for row in focus_rows], "matrix": matrix, "overlap_counts": overlap_counts}

File: src/engine/trades.ts
import { AssetMeta, EventReturns, poiRet, unitLabel } from '@/engine/returns';
import { getLiveReturnPointAtOrBefore, LiveSeriesStateLike } from '@/engine/live';
import { nanMax, nanMean, nanMedian, nanMin, nanPercentile, nanStd, corrcoef } from '@/lib/math';
import { stars, statusFromPctile } from '@/lib/format';

export interface TradeRow {
  lbl: string;
  cls: string;
  ticker: string;
  dir: 'LONG' | 'SHORT';
  med: number;
  mean: number;
  std: number;
  iqr: number;
  stars: string;
  n: number;
  nTotal: number;
  unit: string;
  isRates: boolean;
  hitRate: number;
  sharpe: number;
  sortino: number;
  skew: number;
  worst: number;
  best: number;
  liveGap: number;
  livePctile: number;
  status: string;
  fwdVals: number[];
}

export interface ForwardHorizonStats {
  horizonLabel: string;
  horizonOffset: number;
  q1: number;
  med: number;
  q3: number;
  hit: number;
  sharpe: number;
  sortino: number;
  skew: number;
  worst: number;
  n: number;
}

export interface DotPlotPoint {
  event: string;
  value: number;
}

export interface LiveDeviationSeries {
  offset: number;
  p25: number | null;
  p75: number | null;
  median: number | null;
  live: number | null;
}

export function computeTradeRows(
  labels: string[],
  eventReturns: EventReturns,
  assetMeta: Record<string, AssetMeta>,
  selectedEvents: string[],
  dayN: number,
  fwdDays: number,
  live: LiveSeriesStateLike,
): TradeRow[] {
  const nSel = selectedEvents.length;
  const forwardOffset = dayN + fwdDays;
  if (forwardOffset <= dayN) return [];

  const rows: TradeRow[] = [];

  for (const label of labels) {
    const fwdVals: number[] = [];
    for (const eventName of selectedEvents) {
      const startValue = poiRet(eventReturns, label, eventName, dayN);
      const finishValue = poiRet(eventReturns, label, eventName, forwardOffset);
      if (!Number.isNaN(startValue) && !Number.isNaN(finishValue)) {
        fwdVals.push(finishValue - startValue);
      }
    }
    if (fwdVals.length < 2) continue;

    const med = nanMedian(fwdVals);
    const mean = nanMean(fwdVals);
    const std = nanStd(fwdVals);
    const iqr = nanPercentile(fwdVals, 75) - nanPercentile(fwdVals, 25);
    const isRates = assetMeta[label]?.is_rates_bp || false;
    const unit = unitLabel(assetMeta[label]);

    const hitRate = fwdVals.filter((value) =>
      (med > 0 && value > 0) || (med < 0 && value < 0)
    ).length / fwdVals.length;

    const direction = med >= 0 ? 1 : -1;
    const adjustedValues = fwdVals.map((value) => direction * value);
    const meanAdjusted = nanMean(adjustedValues);
    const stdAdjusted = nanStd(adjustedValues);
    const sharpe = meanAdjusted / (stdAdjusted + 1e-9);
    const downsideValues = adjustedValues.filter((value) => value < 0);
    const downsideStd = downsideValues.length > 1 ? nanStd(downsideValues) : stdAdjusted + 1e-9;
    const sortino = meanAdjusted / (downsideStd + 1e-9);
    const worst = direction > 0 ? nanMin(fwdVals) : nanMax(fwdVals);
    const best = direction > 0 ? nanMax(fwdVals) : nanMin(fwdVals);

    let skew = 0;
    if (fwdVals.length >= 3) {
      const meanValue = nanMean(fwdVals);
      const stdValue = nanStd(fwdVals);
      if (stdValue > 0) {
        skew = fwdVals.reduce((sum, value) => sum + ((value - meanValue) / stdValue) ** 3, 0) / fwdVals.length;
      }
    }

    let liveGap = Number.NaN;
    let livePctile = Number.NaN;
    const livePoint = getLiveReturnPointAtOrBefore(live, label, dayN);
    if (livePoint) {
      const historicalAtLivePoint: number[] = [];
      for (const eventName of selectedEvents) {
        const value = poiRet(eventReturns, label, eventName, livePoint.offset);
        if (!Number.isNaN(value)) historicalAtLivePoint.push(value);
      }
      if (historicalAtLivePoint.length >= 2) {
        liveGap = direction * (livePoint.value - nanMedian(historicalAtLivePoint));
        livePctile = (historicalAtLivePoint.filter((value) => livePoint.value > value).length / historicalAtLivePoint.length) * 100;
      }
    }

    const meta = assetMeta[label] || ({} as AssetMeta);
    rows.push({
      lbl: label,
      cls: meta.class || '',
      ticker: meta.ticker || '',
      dir: med >= 0 ? 'LONG' : 'SHORT',
      med,
      mean,
      std,
      iqr,
      stars: stars(iqr, med),
      n: fwdVals.length,
      nTotal: nSel,
      unit,
      isRates,
      hitRate,
      sharpe,
      sortino,
      skew,
      worst,
      best,
      liveGap,
      livePctile,
      status: statusFromPctile(Number.isNaN(livePctile) ? null : livePctile),
      fwdVals,
    });
  }

  rows.sort((left, right) => right.sharpe - left.sharpe);
  return rows;
}

export function computePerHorizonStats(
  label: string,
  selectedEvents: string[],
  dayN: number,
  eventReturns: EventReturns,
  horizons: Array<{ label: string; offset: number }>,
): ForwardHorizonStats[] {
  const rows: ForwardHorizonStats[] = [];

  for (const horizon of horizons) {
    if (horizon.offset <= dayN) continue;
    const vals: number[] = [];
    for (const eventName of selectedEvents) {
      const start = poiRet(eventReturns, label, eventName, dayN);
      const finish = poiRet(eventReturns, label, eventName, horizon.offset);
      if (!Number.isNaN(start) && !Number.isNaN(finish)) {
        vals.push(finish - start);
      }
    }
    if (vals.length < 2) continue;
    const med = nanMedian(vals);
    const mean = nanMean(vals);
    const std = nanStd(vals);
    const downside = vals.filter((value) => value < 0);
    const downsideStd = downside.length > 1 ? nanStd(downside) : std + 1e-9;
    const dir = med >= 0 ? 1 : -1;
    const directionAdjusted = vals.map((value) => value * dir);
    rows.push({
      horizonLabel: horizon.label,
      horizonOffset: horizon.offset,
      q1: nanPercentile(vals, 25),
      med,
      q3: nanPercentile(vals, 75),
      hit: vals.filter((value) => value * dir > 0).length / vals.length,
      sharpe: (mean * dir) / (std + 1e-9),
      sortino: (mean * dir) / (downsideStd + 1e-9),
      skew: vals.length >= 3 ? vals.reduce((sum, value) => sum + ((value - mean) / (std + 1e-9)) ** 3, 0) / vals.length : 0,
      worst: nanMin(directionAdjusted),
      n: vals.length,
    });
  }

  return rows;
}

export function buildDotPlot(
  label: string,
  selectedEvents: string[],
  startOffset: number,
  endOffset: number,
  eventReturns: EventReturns,
): DotPlotPoint[] {
  const points: DotPlotPoint[] = [];
  if (endOffset <= startOffset) return points;

  for (const eventName of selectedEvents) {
    const start = poiRet(eventReturns, label, eventName, startOffset);
    const finish = poiRet(eventReturns, label, eventName, endOffset);
    if (!Number.isNaN(start) && !Number.isNaN(finish)) {
      points.push({ event: eventName, value: finish - start });
    }
  }
  return points;
}

export function buildLiveDeviationSeries(
  label: string,
  selectedEvents: string[],
  dayN: number,
  eventReturns: EventReturns,
  live: LiveSeriesStateLike,
): LiveDeviationSeries[] {
  const series: LiveDeviationSeries[] = [];

  for (let offset = 0; offset <= dayN; offset += 1) {
    const values: number[] = [];
    for (const eventName of selectedEvents) {
      const value = poiRet(eventReturns, label, eventName, offset);
      if (!Number.isNaN(value)) values.push(value);
    }
    const livePoint = getLiveReturnPointAtOrBefore(live, label, offset);
    series.push({
      offset,
      p25: values.length > 0 ? nanPercentile(values, 25) : null,
      p75: values.length > 0 ? nanPercentile(values, 75) : null,
      median: values.length > 0 ? nanMedian(values) : null,
      live: livePoint?.value ?? null,
    });
  }

  return series;
}

export function buildIdeaCorrelationMatrix(
  rows: TradeRow[],
  selectedEvents: string[],
  dayN: number,
  fwdDays: number,
  eventReturns: EventReturns,
): { labels: string[]; matrix: number[][] } | null {
  const focusRows = rows.slice(0, Math.min(12, rows.length));
  if (focusRows.length < 2) return null;

  const endOffset = dayN + fwdDays;
  const perAssetValues = focusRows.map((row) => {
    return selectedEvents.map((eventName) => {
      const start = poiRet(eventReturns, row.lbl, eventName, dayN);
      const finish = poiRet(eventReturns, row.lbl, eventName, endOffset);
      return !Number.isNaN(start) && !Number.isNaN(finish) ? finish - start : Number.NaN;
    });
  });

  const matrix = focusRows.map((_, i) => {
    return focusRows.map((__, j) => {
      const a = perAssetValues[i];
      const b = perAssetValues[j];
      const filteredA: number[] = [];
      const filteredB: number[] = [];
      for (let index = 0; index < a.length; index += 1) {
        if (!Number.isNaN(a[index]) && !Number.isNaN(b[index])) {
          filteredA.push(a[index]);
          filteredB.push(b[index]);
        }
      }
      if (filteredA.length < 3) return Number.NaN;
      return corrcoef(filteredA, filteredB);
    });
  });

  return {
    labels: focusRows.map((row) => row.lbl),
    matrix,
  };
}
